c++/src/kj/list.h

```diff
@@ -202,6 +203,231 @@ class ListIterator {
   friend class List<T, link>;
 };
 
-} // namespace kj
+template <typename T>
+class OwnList {
+  // A linked list that does its own memory allocation.
+  //
+  // Keep in mind that each new element requires the allocation of two pointers and a T. This is
+  // very little if T is actually an Own<U> but can be quite expensive if T is a large type.
+
+  struct Node {
+    // A simple linked list node which extends the lifetime of the next node.
+
+    Node(T&& value): value{ kj::mv(value) } {}
+
+    Maybe<Own<Node>> maybeNext;
+    Node* prev = nullptr;
+    T value;
+  };
+
+public:
+  template <typename TOrConstT>
+  class Iterator {
+    // A very vanilla pointer-based iterator.
+
+    friend class OwnList;
+
+  public:
+    Iterator(Node* node): node(node) {}
+
+    TOrConstT& operator*() {
+      KJ_IREQUIRE(node != nullptr, "tried to dereference end of list");
+      return node->value;
+    }
+    const T& operator*() const {
+      KJ_IREQUIRE(node != nullptr, "tried to dereference end of list");
+      return node->value;
+    }
+
+    TOrConstT* operator->() {
+      KJ_IREQUIRE(node != nullptr, "tried to dereference end of list");
+      return node->value;
+    }
+    const T* operator->() const {
+      KJ_IREQUIRE(node != nullptr, "tried to dereference end of list");
+      return node->value;
+    }
+
+    Iterator& operator++() {
+      KJ_IREQUIRE(node != nullptr, "tried to advance past end of list");
+
+      KJ_IF_MAYBE (next, node->maybeNext) {
+        node = next->get();
+      } else {
+        node = nullptr;
+      }
+      return *this;
+    }
+    Iterator operator++(int) {
+      auto result = Iterator(node);
+      this->operator++();
+      return result;
+    }
+
+    bool operator==(const Iterator& other) const {
+      return node == other.node;
+    }
+    bool operator!=(const Iterator& other) const {
+      return node != other.node;
+    }
+
+  private:
+    Node* node;
+  };
+
+  using MutableIterator = Iterator<T>;
+  using ConstIterator = Iterator<const T>;
+
+  OwnList() = default;
+  KJ_DISALLOW_COPY(OwnList);
+
+  void pushBack(T v) {
+    // Push a value onto the end of the list.
+    // TODO(now) Should this be add[Back]() instead?
```

Comment:
Hmmmm, I think `add()` + `add(Iterator)` can possible be more within the culture then. `pushFront()` is just `add(begin())`. I think I lean in favor of returning the next iterator from whatever does the erasing. I  think this can be equivalent to `popBack()` and `popFront()` if I implement `remove(Iterator)` and `Iterator::operator--()`. (No idea how strongly you feel about it being distinct from `List::remove()`.)

Reply:
> I think I lean in favor of returning the next iterator from whatever does the erasing.

I definitely understand your preference here -- but I think it's confusing if List and OwnList work differently.

I guess they already work differently due to intrusive vs. non-intrusive.

But... Hmm, maybe we've lost track of the bigger picture here. When is OwnList better than std::deque? It seems like there are two possible advantages:
1. Supporting efficient insertion and deletion mid-list -- but our present use case doesn't need this.
2. Avoiding allocations on insertion and deletion, by relying on an embedded list link (intrusive linked list). But this is not the design being proposed, and our present use case wouldn't benefit from it.
3. Avoiding std in favor of KJ-native data structures. But then, why a new list type, vs. introducing a `kj::Deque`, which would be more efficient?

Maybe we should just use `std::deque` for now and leave it as a TODO to create a `kj::Deque`?

I am worried that if we introduce a non-intrusive `OwnList` now, then it's harder to introduce an intrusive version later.

Sorry to give this feedback so late in the process. :/

Reply:
It is a mite frustrating to hear now but it's a good conversation to be having. I have two points that may or may not be valuable to you:
- There is one more advantage: Traditional linked lists are generally more efficient for small size sequential data storage. This is actually what made me shy away from using a deque here, my use case maintains a list of one in the ideal case.
- If I was writing a `kj::OwnDeque<T>`, I'd write a `kj::OwnList<T>` as well because I would consider it syntactic sugar on top of `kj::OwnList<DequeBlock<T>>`.

Reply:
I think `kj::Deque<T>` would "own" the elements by default, just like `kj::Vector<T>`, `kj::Array<T>`, etc. So there's no need for `kj::OwnDeque<T>`.

Are deques normally implemented using a list of blocks? I was thinking it could be implemented as something like a cross between a vector an a ring buffer...

> Traditional linked lists are generally more efficient for small size sequential data storage.

Hmm, is that because `std::deque` allocates a large initial buffer? `kj::Vector` starts with a 4-element array FWIW (unless you pass something different to the constructor).

Reply:
> Are deques normally implemented using a list of blocks? I was thinking it could be implemented as something like a cross between a vector an a ring buffer...

I must admit I had to go do a bit of research to confirm my understanding. std::deque is a "vector of vectors", not a "list of vectors" like I remembered. The big advantage to that is that you get bounded cost for insertion and removal in the middle of the queue. This flexibility is considered attractive for a general library type.

Ring buffers are a perfectly valid way to implement the core promises of a deque and people do use those sort of deques. The major pain point is the same as most vectors, you can end up with "stop the world" moments when it needs to grow. In my past life in fintech, we used a lot of fixed size ring buffers for queues. Honestly, they work pretty well as long as you think hard about what happens when you hit the limit. I'd be happy to write a ring buffer type instead if you'd prefer that.

Reply:
> Hmm, is that because std::deque allocates a large initial buffer? kj::Vector starts with a 4-element array FWIW (unless you pass something different to the constructor).

IIRC, both `std::deque` and `std::vector` do no allocations until first use. But yeah, deque has a 512B (1/8th a page) block size. That's a sizable reservation that we'll only use if something is wrong.

Reply:
`kj::Vector` also doesn't allocate until first use, but then it allocates an array of length 4.

STL's tendency to disregard small containers in optimization decisions is pretty annoying. (Though I suppose this is an implementation-specific decision. Note that we use libc++, not libstdc++. I wonder if they make the same decisions.)

Is there any way to instruct `std::deque` to reserve a smaller amount of space to start? If so maybe we could just live with that.

IMO vector-of-vectors is overcomplicated. If we really want to implement our own deque, I'd make it a ring buffer on top of a single vector. However, this may still be somewhat complicated to implement as kj::Vector at present is just a thin wrapper around kj::ArrayBuilder which is not designed for the ring buffer scenario. Presumably we want to control constructor and destructor calls so that the unused slots aren't constructed at all, rather than default-initializing them. kj::ArrayBuilder is designed to deal with that. But we probably can't just extend it to support a ring buffer as that'll increase the size and complexity of all ArrayBuilders (and Vectors) to support an obscure use case. So I suppose our Deque type will have to do all its own allocation and constructor management, duplicating a lot of ArrayBuilder functionality.

This would all be useful to have around long-term, but it's quite a yak shave for this ticket. I don't think it's the best thing to spend time on at the moment.

So I think I'd say you should either:
a. Just use whatever STD type makes the most sense.
b. Implement a simple owning linked list, but don't try to make it public for now (so that we can defer the decision on whether it should be intrusive until we have more use cases).

Reply:
Design concerns aside, I think I'll just use a `std::list`. If we see issues, we can switch to `std::deque` or make our own type.